Deprecate functions vectorized via `@vectorize_(1|2)arg` in favor of compact broadcast syntax

Sacha0 opened the pull request:
This PR deprecates all functions vectorized via macros `vectorize_1arg` and `vectorize_2arg` in favor of compact broadcast syntax (ref. #16285). Many non-macro-vectorized functions remain for a separate PR. Should this PR deprecate the `vectorize_1arg` and `vectorize_2arg` macros themselves alongside their usage in base? Best!
- [x] ~~TODO: Vectorized functions involving dates/times.~~ Done.
- [x] ~~TODO: Prettify deprecation warnings. See #17289.~~ Fixed in #17521.
- [x] ~~TODO: Resolve issue with the last test in test/operators.jl. See #17304.~~ Thanks `@stevengj`!
- [x] ~~TODO: Resolve deprecation warnings issued by [markdown's colwidths](https://github.com/JuliaLang/julia/blob/853317d6a22577f336a5cb9f4a031b61516d8ec8/base/markdown/GitHub/table.jl#L81). @MikeInnes, how would you rewrite that line less vectorized `max`? My attempts broke markdown, so I imagine I don't understand the intention. Thanks!~~

Ref. #17265 (devectorization of functions over `SparseMatrixCSC`s).

Sacha0 committed:
Deprecate vectorized functions in base/special/trig.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/special/log.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/special/gamma.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/special/erf.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/special/bessel.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/math.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/floatfuncs.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/fastmath.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/complex.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/dates/accessors.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/dates/adjusters.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/dates/conversions.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/dates/query.jl in favor of compact broadcast syntax.

nalimilan commented:
+100 *if* we are sure this won't affect performance. I guess this depends on `promote_op` methods being correctly defined for all these operations/types...

stevengj commented:
I feel like this is best delayed for 0.6 — it's going to lead to a huge number of deprecation warnings in packages.

JeffBezanson commented:
> I guess this depends on promote_op methods being correctly defined for all these operations/types

Indeed, the `promote_op` approach does not scale at all.

tkelman commented on base/deprecated.jl:
```diff
@@ -788,6 +788,80 @@ function symperm{Tv,Ti}(A::SparseMatrixCSC{Tv,Ti}, pinv::Vector{Ti})
+        :airyprime, :airyai, :airyaiprime, :airybi, :airybiprime, :airy, :airyx, :besselj0, :besselj1, :bessely0, :bessely1, # base/special/bessel.jl
+        :cbrt, :sinh, :cosh, :tanh, :atan, :asinh, :exp, :erf, :erfc, :exp2, :expm1, :exp10, :sin, :cos, :tan, :asin, :acos, :acosh, :atanh, #=:log,=# :log2, :log10, :lgamma, #=:log1p,=# :sqrt, # base/math.jl
+        :abs, :abs2, :angle, :isnan, :isinf, :isfinite, # base/floatfuncs.jl
+        :acos_fast, :acosh_fast, :angle_fast, :asin_fast, :asinh_fast, :atan_fast, :atanh_fast, :cbrt_fast, :cis_fast, :cos_fast, :cosh_fast, :exp10_fast, :exp2_fast, :exp_fast, :expm1_fast, :lgamma_fast, :log10_fast, :log1p_fast, :log2_fast, :log_fast, :sin_fast, :sinh_fast, :sqrt_fast, :tan_fast, :tanh_fast, # base/fastmath.jl
```
I like that these are annotated with where they came from, but should spread over multiple lines when it gets this long

tkelman commented on base/sysimg.jl:
```diff
@@ -294,12 +294,12 @@ include("client.jl")
 
 # dense linear algebra
+include("broadcast.jl")
+importall .Broadcast
```
probably before the linear algebra comment since broadcast is a bit more general than that

Sacha0 commented:
> I feel like this is best delayed for 0.6 — it's going to lead to a huge number of deprecation warnings in packages.

Agreed. This change wait for 0.6 for several reasons. Best!

Sacha0 committed:
Clean up @vectorize_(1|2)arg deprecations.

stevengj commented:
Note that this will close #4363.

Sacha0 commented:
Rebased for 0.6 and passing CI. Nanosoldier? Best!

musm commented:
How does this affect performance? I think a nanosoldier run could be worthwhile

KristofferC commented:
There are very few benchmarks for sparse arrays in `BaseBenchmarks.jl`.

Sacha0 commented:
@KristofferC, most functionality this pull request touches is not immediately related to sparse arrays. Conjecturing the contrary was reasonable though, given my predilections :).

KristofferC commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17302#issuecomment-238655772) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/0a881e8255654c4d55b942c7e22f99ddd56613b5/996d567_vs_008d8e7/report.md). cc @jrevels

Sacha0 commented:
Rebased resolving a deprecation conflict. Prior to rebase, this PR passed both AppVeyor and Travis. Nanosoldier appears happy as well. (As `@pabloferz` pointed out in https://github.com/JuliaLang/julia/pull/17929#issuecomment-238827433, the `lucompletepiv` regressions seems unrelated.) Is this PR in good shape? Thanks!

stevengj commented:
Test failure due to use of `abs(x)` in `test/inference.jl:280`.

stevengj commented:
Might require some manual updates as well?

Sacha0 commented:
> Test failure due to use of abs(x) in test/inference.jl:280.

Introduced by #17847. I'm not certain how to migrate that test to compact broadcast syntax (or otherwise modify it): That test checks that `typeof(abs(Integer[])) == Array{Any,1}` rather than `Array{Union{},1}`. And as `@stevengj` points out in https://github.com/JuliaLang/julia/pull/17847#issuecomment-238744364, `broadcast` does not use the logic the macro-vectorized `abs` call exercises. Hence migrating to compact broadcast syntax would not capture the test's intention (nor would it succeed, as `typeof(abs.(Integer[])) == Arrray{Union{},1}` presently). Thoughts? Thanks!

stevengj commented:
Can you file an issue to extend #17847 to `map` and `broadcast`?

tkelman commented:
I guess it can be done in a separate PR from this one, but I think deprecating the `vectorize_1arg` etc macros themselves would also make sense

stevengj commented:
I agree, @tkelman.  That is, replace all calls to `@vectorize_1arg` in Base with e.g. `@deprecate_vectorize_1arg`, and then add a new definition of `@vectorize_1arg` to `deprecated.jl` that prints a deprecation warning and then calls `@deprecate_vectorize_1arg`.

stevengj commented on base/deprecated.jl:
```diff
@@ -785,4 +785,89 @@ const _oldstyle_array_vcat_ = false
+    S = esc(S)
+    f = esc(f)
+    quote
+        @deprecate $f{T1<:$S,T2<:$S}(x::T1, y::AbstractArray{T2}) $f.(x,y)
```
You can just use `x::$S` here, and similarly in the next line.

Sacha0 replied:
Good catch. Simplified. Thanks!

stevengj commented on base/deprecated.jl:
```diff
@@ -785,4 +785,89 @@ const _oldstyle_array_vcat_ = false
+            :daysofweekinmonth, :monthname, :monthabbr, :daysinmonth, # base/dates/query.jl
+            :isleapyear, :dayofyear, :daysinyear, :quarterofyear, :dayofquarter, # base/dates/query.jl
+    )
+    @eval @dep_vectorize_1arg Dates.TimeType $f
```
Seems like this depends on #16966 — i.e. we need `broadcast` to recognize `TimeType` as a "scalar".

stevengj replied:
Oh, no, I guess it will work as-is for operations on `Array{TimeType}` by themselves.

Sacha0 committed:
Deprecate `@vectorize_1arg` and `@vectorize_2arg`.

Sacha0 commented:
Rebased and fixed a few recently introduced tests involving vectorized functions. Then ...

> I guess it can be done in a separate PR from this one, but I think deprecating the vectorize_1arg etc macros themselves would also make sense

> I agree, @tkelman. That is, you've already replaced all calls to @vectorize_1arg in Base with @dep_vectorize_1arg, and then add a new definition of @vectorize_1arg to deprecated.jl that prints a deprecation warning and then calls @dep_vectorize_1arg.

... added a commit deprecating `@vectorize_1arg` and `@vectorize_2arg`. Getting the deprecated calls to work without complaint and with non-mangled deprecation warnings heightened my appreciation for the objectives of #10940. Best!

Sacha0 committed:
Update documentation following deprecation of macro-vectorized functions and deprecation of `@vectorize_1arg` and `@vectorize_2arg` themselves. Also add cross-referenes between max and maximum, and also between min and minimum.

Sacha0 commented:
> Might require some manual updates as well?

I've added a commit that: (1) removes mentions of operation over arrays from the documentation for the formerly-macro-vectorized functions; and (2) also removes the manual subsection on `@vectorize_1arg`/`@vectorize_2arg`. Other documentation modifications that should accompany this PR? Best!

Sacha0 commented:
With benefit of #18200 this passes all tests again. Anything I've missed? Thanks!

stevengj commented on base/dates/conversions.jl:
```diff
@@ -131,11 +128,4 @@ end
-@vectorize_1arg DateTime datetime2rata
-@vectorize_1arg Real julian2datetime
-@vectorize_1arg DateTime datetime2julian
+datetime2julian(dt::DateTime) = (value(dt) - JULIANEPOCH)/86400000.0
```
missing newline

stevengj commented on base/sparse/sparsematrix.jl:
```diff
@@ -295,7 +295,7 @@ Convert a sparse matrix or vector `S` into a dense matrix or vector.
 full
 
-float(S::SparseMatrixCSC) = SparseMatrixCSC(S.m, S.n, copy(S.colptr), copy(S.rowval), float(copy(S.nzval)))
+float(S::SparseMatrixCSC) = SparseMatrixCSC(S.m, S.n, copy(S.colptr), copy(S.rowval), float.(copy(S.nzval)))
```
Maybe just `float.(S.nzval)` here?   Calling `copy` seems to result in two copies being made.

stevengj commented on base/special/erf.jl:
```diff
@@ -216,6 +214,4 @@ function erfcinv(y::Float32)
-erfcinv(x::Integer) = erfcinv(float(x))
-@vectorize_1arg Real erfcinv
-
+erfcinv(x::Integer) = erfcinv(float(x))
```
missing newline

stevengj commented on doc/manual/arrays.rst:
```diff
@@ -474,56 +474,8 @@ operators ``/`` and ``\`` operate elementwise when the denominator is a scalar.
-    2×3 Array{Int64,2}:
-      1   4  16
-     25  36  49
+To enable convenient vectorization of other operations, Julia provides compact syntax
```
```
To enable convenient vectorization of mathematical and other operations
Julia provides the compact syntax `f.(args...)`, e.g. `sin.(x)` or `min.(x,y)`, for elementwise operations
over arrays or mixtures of arrays and scalars (a :func:`broadcast`) operation.  See
See :ref:`man-dot-vectorizing`:.
```

Sacha0 commented:
Newlines inserted, `float.(copy(S.nzval))` reduced to `float.(S.nzval)`, and documentation revised as indicated. Much thanks for the review!

nalimilan commented on base/deprecated.jl:
```diff
@@ -865,4 +865,118 @@ for f in (:sin, :sinh, :sind, :asin, :asinh, :asind,
+for f in (
+        :polygamma, :zeta, :beta, :lbeta, # base/special/gamma.jl
+        :airy, :airyx, :besseli, :besselix, :besselj, :besseljx, # base/special/bessel.jl
+            :besselk, :besselkx, :bessely, :besselyx, :besselh, # base/special/bessel.jl
```
Indentation issue here.

Sacha0 replied:
I intended the indentation level of that and similar lines in base/deprecated.jl to reflect continuation of the list on the preceding line. (Notice that, e.g., that group of three lines lists functions from the same file.) Happy to change that if you prefer though? Thanks!

nalimilan replied:
What I mean is that the third line has one more level of indentation than the two first.

Sacha0 replied:
Yes, exactly :). The third line (beginning `:besselk`, thrice indented) is a continuation of the second line's list (beginning `:airy`, twice indented), the functions on the second and third lines coming from the same file (base/special/bessel.jl per the trailing comments). The same extra-indentation-for-list-continuation occurs in the earlier loops over functions as well, reflecting the file grouping of functions. If you prefer removal of the extra indentation indicating file grouping, I am happy to change that uniformly. Thanks!

yuyichao replied:
I think a better way to show grouping that's compatible with indent rules is to use the comment to separate groups.

Sacha0 replied:
In the loops involving multi-line lists, the file-comments now separate those lists as suggested. How does that look? Thanks!

nalimilan replied:
Looks good, thanks!

nalimilan commented on doc/manual/arrays.rst:
```diff
@@ -474,56 +474,10 @@ operators ``/`` and ``\`` operate elementwise when the denominator is a scalar.
+To enable convenient vectorization of mathematical and other operations, Julia provides
+the compact syntax ``f.(args...)``, e.g. ``sin.(x)`` or ``min.(x,y)``, for elementwise
+operations over arrays or mixtures of arrays and scalars (a :func:`broadcast` operation).
+See :ref:`man-dot-vectorizing`:.
```
The final `:` shouldn't be there, right?

stevengj replied:
Right, no final `:`.

Sacha0 replied:
Good catch. Fixed. Thanks!

stevengj commented on base/linalg/arpack.jl:
```diff
@@ -114,9 +114,9 @@ function eupd_wrapper(T, n::Integer, sym::Bool, cmplx::Bool, bmat::String,
+    dmap = x->abs.(x)
     if iparam[7] == 3 # shift-and-invert
-        dmap = x->abs(1./(x-sigma))
+        dmap = x->abs.(1./(x-sigma))
```
Might as well use `x.-sigma` here so that everything gets fused in 0.6

stevengj commented on base/linalg/diagonal.jl:
```diff
@@ -231,17 +231,17 @@ ctranspose(D::Diagonal) = conj(D)
 trace(D::Diagonal) = sum(D.diag)
 det(D::Diagonal) = prod(D.diag)
-logdet{T<:Real}(D::Diagonal{T}) = sum(log(D.diag))
+logdet{T<:Real}(D::Diagonal{T}) = sum(log.(D.diag))
```
`sum(log, D.diag)` will avoid the temporary array completely.

stevengj replied:
Similarly below.

stevengj commented on base/linalg/diagonal.jl:
```diff
@@ -285,7 +285,7 @@ function pinv{T}(D::Diagonal{T})
 function pinv{T}(D::Diagonal{T}, tol::Real)
     Di = similar(D.diag, typeof(inv(zero(T))))
-    if( !isempty(D.diag) ) maxabsD = maximum(abs(D.diag)) end
+    if( !isempty(D.diag) ) maxabsD = maximum(abs.(D.diag)) end
```
(We should really have a `maximum(::Function, iterable)` method analogous to `sum`.  But that's a task for another PR.)

Sacha0 replied:
Now #18310

stevengj commented on base/linalg/diagonal.jl:
```diff
@@ -303,10 +303,10 @@ eigvecs(D::Diagonal) = eye(D)
 
 #Singular system
-svdvals{T<:Number}(D::Diagonal{T}) = sort(abs(D.diag), rev = true)
+svdvals{T<:Number}(D::Diagonal{T}) = sort(abs.(D.diag), rev = true)
```
`sort!` would eliminate a temporary

stevengj commented on base/linalg/generic.jl:
```diff
@@ -372,10 +372,10 @@ cond(x::Number) = x == 0 ? Inf : 1.0
-condskeel(A::AbstractMatrix, x::AbstractVector, p::Real=Inf) = norm(abs(inv(A))*abs(A)*abs(x), p)
-condskeel{T<:Integer}(A::AbstractMatrix{T}, x::AbstractVector, p::Real=Inf) = norm(abs(inv(float(A)))*abs(A)*abs(x), p)
+condskeel(A::AbstractMatrix, p::Real=Inf) = norm(abs.(inv(A))*abs.(A), p)
+condskeel{T<:Integer}(A::AbstractMatrix{T}, p::Real=Inf) = norm(abs.(inv(float(A)))*abs.(A), p)
```
Why is this method necessary?  `inv(A)` already converts to float for integer arrays, so it seems like we could use the generic method above.

Sacha0 replied:
Good question. Perhaps @jiahao could provide some insight, having added the integer-specialized methods alongside tests in https://github.com/JuliaLang/julia/commit/797757103ee2b6d235763de76371335a10558088 ?

jiahao replied:
The integer-valued matrix case used to throw errors. If they don't anymore we can get rid of them.

Sacha0 replied:
Added a commit removing the integer-specialized `condskeel` methods. What say you, CI?

stevengj commented on base/linalg/generic.jl:
```diff
@@ -372,10 +372,10 @@ cond(x::Number) = x == 0 ? Inf : 1.0
-condskeel{T<:Integer}(A::AbstractMatrix{T}, x::AbstractVector, p::Real=Inf) = norm(abs(inv(float(A)))*abs(A)*abs(x), p)
+condskeel(A::AbstractMatrix, p::Real=Inf) = norm(abs.(inv(A))*abs.(A), p)
+condskeel{T<:Integer}(A::AbstractMatrix{T}, p::Real=Inf) = norm(abs.(inv(float(A)))*abs.(A), p)
+condskeel(A::AbstractMatrix, x::AbstractVector, p::Real=Inf) = norm(abs.(inv(A))*abs.(A)*abs.(x), p)
```
This should be parenthesized `(abs.(A)*abs.(x))` in order to multiply matrix*vector rather than matrix*matrix.

stevengj commented on base/linalg/symmetric.jl:
```diff
@@ -266,9 +266,9 @@ for (funm, func) in ([:logm,:log], [:sqrtm,:sqrt])
+                retmat = (F.vectors * Diagonal(($func).(F.values))) * F.vectors'
             else
-                retmat = (F.vectors * Diagonal(($func)(complex(F.values)))) * F.vectors'
+                retmat = (F.vectors * Diagonal(($func).(complex(F.values)))) * F.vectors'
```
Can't we now do `($func).(complex.(F.values))`, and this will be better anyway because it will fuse the loop?

Should we get rid of `float` and `complex` for arrays in favor of dot calls?

Sacha0 replied:
> Shouldn't we deprecate `float` and `complex` and `big` for arrays in favor of dot calls?

Absolutely. This PR attacks only macro-vectorized methods, leaving manually vectorized methods for a followup PR. (My approach has been to select a macro-vectorized method, deprecate it, find its usages via depwarns, migrate those usages, rinse and repeat. As such I presently miss, e.g., `float` calls that dispatch to manually vectorized methods. But I've been accumulating a list of manually vectorized functions to attack in the followup PR and much look forward to diving in.)

stevengj commented on base/markdown/GitHub/table.jl:
```diff
@@ -78,7 +78,7 @@ end
 
 colwidths(rows; len = length, min = 0) =
-    reduce(max, [min; convert(Vector{Vector{Int}}, mapmap(len, rows))])
+    reduce((x,y) -> max.(x,y), [min; convert(Vector{Vector{Int}}, mapmap(len, rows))])
```
Can't we just call `maximum` here?

Sacha0 replied:
My thought as well. But that replacement caused test failures (see the original post's checklist's last item). I arrived at the present solution after examining that line more thoroughly. The details escape me due to the intervening time.

stevengj commented on doc/manual/arrays.rst:
```diff
@@ -474,56 +474,10 @@ operators ``/`` and ``\`` operate elementwise when the denominator is a scalar.
-
-Note that there is a difference between :func:`min` and :func:`max`, which operate
-elementwise over multiple array arguments, and :func:`minimum` and :func:`maximum`, which
-find the smallest and largest values within an array.
```
Note that we should still document this difference.

Sacha0 replied:
Documentation sketch added [here](https://github.com/JuliaLang/julia/pull/17302/files#diff-3794b546e80ba572c70876d9cf49112aR680). Is that roughly what you had in mind? I bet there is a more appropriate location? Thanks!

stevengj replied:
No, I don't think specific information on `max` belongs in that section on dot vectorization syntax.  Just leaving it in the same place in `arrays.rst` should be fine for now, without an example.

(The `max` and `maximum` docstrings should cross-reference each other, too, if they aren't already doing so.)

Sacha0 replied:
Moved the sketch to `arrays.rst`, nixed the example, and added cross references between `max` (`min`) and `maximum` (`minimum`). Thanks!

stevengj commented:
(One of the side effects of doing this sort of thing is that it forces us to proofread all of the vectorized-math calls, hence the discovery of somewhat unrelated things that should be cleaned up.  Thanks for working on this, @Sacha0!)

Sacha0 committed:
Revise a few vectorized-math methods, specifically eliminating temporaries and enabling broadcast loop fusion.

Sacha0 committed:
Remove now-redundant integer-specialized condskeel methods.

Sacha0 commented:
> (One of the side effects of doing this sort of thing is that it forces us to proofread all of the vectorized-math calls, hence the discovery of somewhat unrelated things that should be cleaned up. Thanks for working on this, @Sacha0!)

A pleasure, and I tremendously appreciate the thorough review! Best!

tkelman commented on doc/manual/functions.rst:
```diff
@@ -677,6 +677,24 @@ will be fused with other nested "dot" calls.  ``X .+= Y`` is equivalent
 to ``X .= X .+ Y`` and will eventually result in a fused in-place assignment.
 Similarly for ``.*=`` etcetera.)
 
+Note that there is a difference between ``max.(a,b)``, which ``broadcast``s :func:`max`
```
I think you need a space (maybe backslash-escaped?) between the trailing ``` `` ``` and the s

Sacha0 replied:
Good catch. Added a `\{space}` after the double backtick (hopefully that's correct). Thanks!

stevengj commented on doc/manual/arrays.rst:
```diff
@@ -474,56 +474,14 @@ operators ``/`` and ``\`` operate elementwise when the denominator is a scalar.
-    2×3 Array{Int64,2}:
-      1   4  16
-     25  36  49
+Note that there is a difference between ``max.(a,b)``, which ``broadcast``\ s :func:`max`
```
This paragraph needs to after the "To enable convenient vectorization" paragraph, which explains the dot-call syntax.

stevengj commented on base/docs/helpdb/Base.jl:
```diff
@@ -1163,7 +1163,7 @@ randsubseq!
     maximum(A, dims)
 
-Compute the maximum value of an array over the given dimensions.
+Compute the maximum value of an array over the given dimensions. Also see :func:`max`.
```
This needs to be in Markdown link syntax, and should probably give some explanation
```
See also the [`max(a,b)`][:func:`max`] function to take the maximum of
two or more arguments, which can be applied elementwise to arrays via `max.(a,b)`.
```

stevengj commented on base/docs/helpdb/Base.jl:
```diff
@@ -2961,7 +2961,7 @@ rank
     max(x, y, ...)
 
-Return the maximum of the arguments. Operates elementwise over arrays.
+Return the maximum of the arguments. Also see :func:`maximum`.
```
Similar to above.
```
See also the [`maximum`][:func:`maximum`] function to take the maximum of
all the elements of an array.
```

Sacha0 commented:
Good catches. Revised. Thanks!

stevengj commented:
LGTM.

tkelman commented:
since this touches docs, let's hold off on merging until after #18041

tkelman commented:
Big doc merge done. Rebase?

Sacha0 commented:
Rebased. Thanks for the ping!
